### DISCUSSION TITLE
.refine change schema type

### DISCUSSION THREAD

**Discussion by Encuba** [colinhacks/zod#3714] [Q&A]
I'm trying to validate a schema with Zod that has an avatar property of type File. I want to set it to null by default so I can easily validate whether there is an avatar or not. My schema looks like this:
```
const personSchema = z.object({
      personID: z.number().optional(),
      firstName: z.string().min(2),
      middleName: z.string().min(2),
      lastNames: z.string().min(2),
      dni: z.string(),
      avatar: z
        .instanceof(File)
        .nullable()
        .refine((file) => file != null, "The avatar is required")
    })
```
However, the problem is that even though I use .nullable(), the .refine() method changes the avatar type to File instead of File | null. This prevents me from setting null as a default value:
```
 type  PersonSchema = z.infer<typeof personSchema>
  const person: PersonSchema = {
    firstName: "",
    middleName: "",
    lastNames: "",
    dni: "",
    avatar: null
```
is there any way to achieve this?

**Comment by puku0x** [comment]
Same here. I guess it is caused by the one of the latest TypeScript features.
https://devblogs.microsoft.com/typescript/announcing-typescript-5-5/#inferred-type-predicates

Avoiding non-null inference may help you.

```ts
const personSchema = z.object({
  avatar: z
    .instanceof(File)
    .nullable()
    .refine((file) => (file !== null ? true : false), 'The avatar is required'),
});
```

```ts
const personSchema = z
  .object({
    avatar: z.instanceof(File).nullable(),
  })
  .superRefine((data, ctx) => {
    if (data.avatar === null) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'The avatar is required',
        path: ['avatar'],
      });
    }
  });
```
